File: go/proverb/proverb.go
package proverb

import "fmt"

func Proverb(rhyme []string) (result []string) {
	if len(rhyme) > 0 {
		for i := 0; i < len(rhyme)-1; i++ {
			result = append(result, fmt.Sprintf("For want of a %v the %v was lost.", rhyme[i], rhyme[i+1]))
		}
		result = append(result, fmt.Sprintf("And all for the want of a %v.", rhyme[0]))
	}
	return
}

File: php/proverb/Proverb.php
<?php

class Proverb {
    public function recite($pieces) {
        $res = [];
        for ($i = 1; $i < count($pieces); ++$i)
            $res[]= sprintf("For want of a %s the %s was lost.", $pieces[$i - 1], $pieces[$i]);

        if ($pieces)
            $res[]= sprintf("And all for the want of a %s.", $pieces[0]);

        return $res;
    }
}
